### ISSUE TITLE
Document content prior to first "Heading 1" paragraph is not exported to ePub

### ISSUE THREAD

**Issue by jhardin-impsec** [eLAIX/eLAIX#35]
No document content prior to the first paragraph having style "Heading 1" is exported to the ePub document. This makes it difficult to create a title page for the document.

This is occurring with eLAIX-4.1.0-2014-04-14.oxt

**Comment by lukekendall** [comment]
I agree: it also means that any copyright text or disclaimers are omitted.  It's a serious problem in my opinion.  This is occurring with eLAIX 4.1 and LibreOffice 4.2.7.2.

**Comment by harry-boldt** [comment]
This effect is due to historical reasons. eLAIX first has been developed to export ILIAS learning modules, that do always start with the first chapter. eBooks however can have important content before the first chapter, as you mentioned. I do agree. So I'm here at a point, that could force me, to separate eLAIX into two different projects. Or even into three projects. Most of the toolbar buttons do help users in creating well formed documents, but they are not necessary neither for the ebook export nor for the ILIAS export. They just make life easier, when creating documents. Already tried to properly arrange images or to create a link to a chapter in Writer? With the eLAIX toolbar button, this can be done in a second. The export walks through the document, trying to understand, what's going on. There are no special tricks (beside the accordion function), that the eLAIX related functions do implement. You can manage all your arrangements with the standard Writer functions as well.

**Comment by lukekendall** [comment]
I understand what you're saying.  Forking the code into two (or three) 
variants would make extra work for you, and then you'd also have the 
problems of fixing many bugs in each version.

What about, since it knows whether it is being used to export ILIAS or 
ePUB, in the cases where there is a difference in functionality, either 
gather the extra data needed always, but only use it where appropriate 
(e.g. in the ePUB case, gathering the cover image and any pre-Heading 1 
text into a data structure), and then on export, use the data when 
required for the format?

I was just doing some experiments with Writer2HTML's epub conversion 
results; I had planned to send you the .epub file it produced, since the 
cover image is included correctly: but although it *is* included in the 
full book, when I try my subset book test file, the image does not get 
included.  Since it's after 2am here, I think I'll give up on the 
problem for today.  I started investigating, because I thought I could 
always unzip the .epub and manually alter the contents to include the 
cover image and preliminary text, as an option, so I started comparing 
the two .epubs produced (eLAIX and Writer2HTML), and the Writer2HTML 
does have content that explains why it has a cover image when I use it 
on my full book, not the test subset version I sent you).

Anyway, enough for today, I think.

Regards,

luke

**Comment by harry-boldt** [comment]
Ups, I did not receive a test subset version yet. Did you realize, that you can define a cover picture in the export dialog? I never encountered any problems with this function till now. So I'm awaiting your test document...

**Comment by jhardin-impsec** [comment]
It's reasonable that the two different targets should have different behaviors. Yet I hope that this doesn't rise to the level of requiring a fork of the code.

Harry, should your comments be interpreted as saying that it would be very difficult to implement export of content prior to the first Heading 1 paragraph? If so, then forking might be a reasonable response. Otherwise, may I echo the suggestion that pre-Heading 1 export should either be intelligent (i.e. look at the export format to decide whether or not to include that content), or be based on a checkbox in the export options dialog?

**Comment by harry-boldt** [comment]
Yes, you're right. The problem here is the technical main concept behind eLAIX: Before the export a document is first completely scanned into a RAM based database with all needed informations not only concerning the text and its formats, but also concerning internal links, footnotes, graphics and so on. On this level, eLAIX doesn't care about wether the export goes to ILIAS or to ePUB. Reason for that: One module that fits all ;-). As long as there are not so many differences, it was okay for me to read some special ePUB related or ILIAS related stuff. But the more we have different tasks the more I'm thinking about forking the project. Of course, forking means fixing a bug twice, what is certainly a disadvantage, as luke already mentioned.

**Comment by jhardin-impsec** [comment]
Would it be feasible to read the content preceding the first Heading 1 paragraph into a "chapter zero" entry in that internal database (i.e. start capturing content at the beginning of the document, rather than at the first Heading 1), and then decide whether or not to export chapter zero (without a heading) or totally ignore it based on the export type or a UI checkbox?

**Comment by harry-boldt** [comment]
This is exactly the point. It could be done this way, but that also means, to walk through unneeded jobs in some cases. Since eLAIX knows from the beginning, which kind of export has to be done, a different way of reading the document would be more performant. A common reading of the document however is the main reason for keeping the project togehter.

**Comment by GreeneDragon** [comment]
I have just realized this problem. The title page and copyright page are not exported into epub. Is there a way to work around this issue?

**Comment by harry-boldt** [comment]
I'm sorry, not yet. But I'm aware of that ;-).

Regards
Harry


--
BOLDT MEDIA
Dr. phil., Dipl.-Ing. Harry Boldt
Medienpädagoge, Ingenieur für Biomedizinische Technik

Eckenberger Straße 28
52066 Aachen
Tel +49 241 46361877
Fax +49 241 46371531
Mobil +49 1520 1901633
Web: www.boldt-media.de
Mail: harry.boldt@boldt-media.de

**Comment by jhardin-impsec** [comment]
It lives!

Re "that also means, to walk through unneeded jobs in some cases": I'd suggest that it is preferable to collect data that in some use cases you ignore, than to ignore data that in some use cases is very important.

I don't see "we don't use that data in ILIAS export" as a reasonable argument for not collecting it at all. Would the performance impact from capturing a small number of extra pages for "chapter zero" really be all that great? Am I missing something critical by using the "chapter zero" mental model for thinking about how to achieve this? That seems to me to be the simplest way to achieve this, but I have no familiarity with how eLAIX is actually designed.
